### PR TITLE
warn code pasted

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 1]
```diff
@@ -541,6 +542,31 @@ export const EditorUI = (props: EditorUIProps) => {
       }
     })
 
+    editor.onDidPaste((e) => {
+      if (e && e.range && e.range.startLineNumber >= 0 && e.range.endLineNumber >= 0 && e.range.endLineNumber - e.range.startLineNumber > 10) {
+        const modalContent: AlertModal = {
```
Comment (ioedeveloper) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:547]: Can this message be displayed once? So it doesn't get annoying to a user.

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 2]
```diff
@@ -541,6 +542,31 @@ export const EditorUI = (props: EditorUIProps) => {
...
+          id: 'newCodePasted',
+          title: 'New pasted code detected!',
+          message: (
+            <div> <i className="fas fa-exclamation-triangle text-danger mr-1"></i>
+              Remix detected you have just pasted a consequent code snippet or contract in the editor.
+              <div>
+                Please make sure you fully understand this new code before executing <b>any</b> transaction that uses it. 
+                <div>
+                Your wallet might be exposed and at risk if you run untrusted code. In a worst-case scenario you could loose all your money.
```
Comment (ioedeveloper) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:556]: It may be good to highlight sensitive parts of the message e.g `risk`, `losing money`, etc.

**Comment by ioedeveloper** [comment]
@yann300 Would it not be better to prevent pasting from happening in a file until the user has dismissed the modal (`Paste` or `Cancel`)

**Comment by yann300** [comment]
> Would it not be better to prevent pasting from happening in a file until the user has dismissed the modal (Paste or Cancel)?

I didn't find any `onWillPaste` event..

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx (hunk 2 above)
Comment (ryestew) [reply libs/remix-ui/editor/src/lib/remix-ui-editor.tsx]: The language needs a slight update:

You have just pasted a code snippet or contract in the editor.
Make sure you fully understand this new code before executing any transaction that uses it.
Your wallet might be exposed and at risk if you run untrusted code. In a worst-case scenario, you could loose all your money.
If you don't fully understand it, please don't run this code.
If you are not a smart contract developer, please ask assistance from someone you trust and has the appropriate skills to determine whether this code is safe to use.

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 3]
```diff
@@ -541,6 +543,32 @@ export const EditorUI = (props: EditorUIProps) => {
       }
     })
 
+    editor.onDidPaste((e) => {
+       if (!pasteCodeRef.current && e && e.range && e.range.startLineNumber >= 0 && e.range.endLineNumber >= 0 && e.range.endLineNumber - e.range.startLineNumber > 10) {
+        const modalContent: AlertModal = {
+          id: 'newCodePasted',
+          title: 'New pasted code detected!',
```
Comment (ryestew) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:550]: title: 'Slow down, Cowboy!  New pasted code detected.

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 4]
```diff
@@ -541,6 +543,32 @@ export const EditorUI = (props: EditorUIProps) => {
...
+          message: (
+            <div> <i className="fas fa-exclamation-triangle text-danger mr-1"></i>
+              Remix detected you have just pasted a consequent code snippet or contract in the editor.
```
Comment (ryestew) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:553]: You have just pasted a code snippet or contract in the editor.

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 5]
```diff
@@ -541,6 +543,32 @@ export const EditorUI = (props: EditorUIProps) => {
...
+              <div>
+                Please make sure you fully understand this new code before executing <b>any</b> transaction that uses it.
```
Comment (ryestew) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:555]: Make sure you fully understand this new code before executing <b>any</b> transaction that uses it.

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 6]
```diff
@@ -541,6 +543,32 @@ export const EditorUI = (props: EditorUIProps) => {
...
+        const modalContent: AlertModal = {
+          id: 'newCodePasted',
+          title: 'New pasted code detected!',
+          message: (
+            <div> <i className="fas fa-exclamation-triangle text-danger mr-1"></i>
+              Remix detected you have just pasted a consequent code snippet or contract in the editor.
+              <div>
+                Please make sure you fully understand this new code before executing <b>any</b> transaction that uses it. 
+                <div>
+                Your wallet might be exposed and <span className='text-warning'> at risk </span> if you run untrusted code. In a worst-case scenario you could <span className='text-warning'>loose all your money</span>.
```
Comment (ryestew) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:557]: Your wallet might be exposed and <span className='text-warning'>at risk</span> if you run untrusted code. In a worst-case scenario, you could <span className='text-warning'>loose all your money</span>.

Comment (ryestew) [reply libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:557]: add a link to our docs about security: https://remix-ide.readthedocs.io/en/latest/security.html

**Review by ryestew — CHANGES_REQUESTED** [review]

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 7]
```diff
@@ -541,6 +543,32 @@ export const EditorUI = (props: EditorUIProps) => {
...
+                </div>
+                <span className='text-warning'>If you don't fully understand it, please don't run this code.</span>
+                <div>
+                If you are not a smart contract developer, please ask assistance to someone you trust and has the appropirate skills to determine whether this code is safe to use.
```
Comment (Aniket-Engg) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:561]: typo in `appropirate `

**Review by Aniket-Engg — COMMENTED** [review]
In case to define whether we should it on each code paste, we can add an option is settings which is by default enabled.

**Comment by yann300** [comment]
> In case to define whether it should be shown on each code paste, we can add an option is settings which is by default enabled.

currently, the message will show one time per session..

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx (hunk 1 above)
Comment (yann300) [reply libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:548]: done

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 8]
```diff
@@ -541,6 +543,33 @@ export const EditorUI = (props: EditorUIProps) => {
...
+              <div>
+                Make sure you fully understand this new code before executing <b>any</b> transaction that uses it.
+                <div>
+                Your wallet might be exposed and <span className='text-warning'> at risk </span> if you run untrusted code. In a worst-case scenario, you could <span className='text-warning'>loose all your money</span>.
+                </div>
+                <span className='text-warning'>If you don't fully understand it, please don't run this code.</span>
+                <div>
+                If you are not a smart contract developer, please ask assistance to someone you trust and has the appropriate skills to determine whether this code is safe to use.
+                </div>
+                <div>see <a href='https://remix-ide.readthedocs.io/en/latest/security.html'> this blog post </a> for more information.</div>
```
Comment (Aniket-Engg) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:563]: It says `this blog post` but actually that takes to a section of Remix docs. It opens the link in the same tab which closes Remix IDE.

`S` in `see` should be in capital letter.

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 9]
```diff
@@ -541,6 +543,33 @@ export const EditorUI = (props: EditorUIProps) => {
...
+    editor.onDidPaste((e) => {
+       if (!pasteCodeRef.current && e && e.range && e.range.startLineNumber >= 0 && e.range.endLineNumber >= 0 && e.range.endLineNumber - e.range.startLineNumber > 10) {
+        const modalContent: AlertModal = {
+          id: 'newCodePasted',
+          title: 'Slow down, Cowboy! New pasted code detected.',
+          message: (
+            <div> <i className="fas fa-exclamation-triangle text-danger mr-1"></i>
+              You have just pasted a code snippet or contract in the editor.
+              <div>
+                Make sure you fully understand this new code before executing <b>any</b> transaction that uses it.
```
Comment (Aniket-Engg) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:555]: `before executing any transaction` can be considered to be updated as `before deploying or interacting with it`

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 10]
```diff
@@ -541,6 +543,33 @@ export const EditorUI = (props: EditorUIProps) => {
...
+    editor.onDidPaste((e) => {
+       if (!pasteCodeRef.current && e && e.range && e.range.startLineNumber >= 0 && e.range.endLineNumber >= 0 && e.range.endLineNumber - e.range.startLineNumber > 10) {
+        const modalContent: AlertModal = {
+          id: 'newCodePasted',
+          title: 'Pasted code Alert',
+          message: (
+            <div> <i className="fas fa-exclamation-triangle text-danger mr-1"></i>
+              You have just pasted a code snippet or contract in the editor.
+              <div>
+                Make sure you fully understand this new code before before deploying or interacting with it.
```
Comment (Aniket-Engg) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:555]: `before` is written twice

**Review by Aniket-Engg — APPROVED** [review]

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 11]
```diff
@@ -541,6 +543,33 @@ export const EditorUI = (props: EditorUIProps) => {
...
+        const modalContent: AlertModal = {
+          id: 'newCodePasted',
+          title: 'Pasted code Alert',
+          message: (
+            <div> <i className="fas fa-exclamation-triangle text-danger mr-1"></i>
+              You have just pasted a code snippet or contract in the editor.
+              <div>
+                Make sure you fully understand this new code before deploying or interacting with it.
+                <div>
+                Your wallet might be exposed and <span className='text-warning'> at risk </span> if you run untrusted code. In a worst-case scenario, you could <span className='text-warning'>loose all your money</span>.
```
Comment (ryestew) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:557]: Here's an edit - but leave the text-warning spans.
Running untrusted code can put your wallet at risk. In a worst-case scenario, you could loose all your money.

**Review by ryestew — CHANGES_REQUESTED** [review]

**Review by ryestew — APPROVED** [review]

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx (hunk 1 above)
Comment (vermouth22) [reply libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:548]: There is a spelling error in the warning message:
"...could loose all your money" should be
"...could LOSE all your money"

**Pull request merged**